rfcs/0107-version-normalization.md

```diff
@@ -0,0 +1,182 @@
+builtins.parseDrvName "mpv-unstable-2021-05-03"
+{ name = "mpv-unstable"; version = "2021-05-03"; }
+```
+
+It happens because the `version` attribute returned by `builtins.parseDrvName`
+starts with a digit. This is not a bug, but a deliberate design decision of Nix
+specification; therefore we should strive to follow it, neither circumventing
+nor ignoring it.
+
+That being said, `version` attribute should start with a digit.
+
+Furthermore, `version` attribute should follow the semantics for upgrading as
+stated in `nix-env` manpage.
+
+This document describes a format suitable to fix this issue while striving for
+simplicity.
+
+# Detailed design
+[design]: #detailed-design
+
+## Terminology
+
+_Disclaimer_: whereas some of the terms enumerated below are borrowed from
+[pkgsrc guide](https://www.netbsd.org/docs/pkgsrc/) plus a bit of terminology
+employed by git, this document aims to be general, not being overly attached to
+them.
+
+- Program denotes the piece of software to be fetched and processed via Nixpkgs.
+  - This term makes no distinction about the immediate or intended uses of the
+    program; it can range from non-interactive programs to full GUI
+    applications, even including filesets intended as input to other programs
+    (such as firmwares and fonts).
+- Team denotes the maintainers of the program.
+  - This term makes no distinction among the various models of organization of a
+    team; it ranges from a solitary programmer to a business company - and
+    everything inbetween.
+- Source denotes the origin of the program.
+  - This term makes no distinction among precompiled, binary-only or high-level
+    code distributions.
+- Snapshot denotes the source of the program taken at a point of the time.
+  - Labeled snapshot denotes any snapshot explicitly labeled by its team.
+  - Unlabeled snapshot denotes any snapshot not explicitly labeled by its team.
+- Release denotes any distributed snapshot, as defined by its team.
+- Branch denotes a logical sequence of snapshots, as identified by the program's
+  team;
+  - Usually these branches are denoted by names such like `stable`, `master`,
+    `unstable`, `trunk`, `experimental`, `staging` etc.;
+
+## Design
+
+- For a labeled snapshot:
+  - `version` should be constituted of the version of the snapshot, as defined
+    by the upstream project, without any alphabetical characters (e.g. "v",
+    "rel") prepending it.
+    - Alphabetical characters after the first should be maintained, except
+      optionally those clearly used as separators, in which case they are
+      replaced by dots (emulating a typical semver).
+
+- For an unlabeled snapshot:
+  - `version` should be constituted of a concatenation of the elements below in
+    this order:
+      - the version of the latest labeled snapshot, as defined above;
+        - If the project never released a labeled snapshot, `0_0` should be used
+          as default.
```

Comment:
How does the change from `0_0` to `i.j` impact lexicographical ordering?
I'm thinking most (all?) systems probably use ASCII values to sort, but it'd be good to both research this and be explicit about it.

Reply:
The lexicographical ordering is the same as used by Nix itself, being registered in the [nix-env manpage](https://www.mankier.com/1/nix-env#Operation_--upgrade-Versions).

I am treating it as previous knowledge here.

The idea of `0_0` is that, being an uncommon numbering scheme (underscores are not employed in Nixpkgs version attrs, as far as I remember), it will not clash anything.

Reply:
I was wondering about systems other than Nix as well, as other programs operating with nixpkgs data may be using the version attribute for their own purposes.

Reply:
From what I see in the nix repl `0_0` sorts as greater ('newer') than for example `0.1`.

Reply:
I am also against `0_0`. It is very intuitive for any reader. If we need to put a number to satisfy `parseDrvName` I think we should just use `0` or something more obvious like `0unreleased`

Reply:
About this, I am pasting the relevant part:

> The versions are compared  by splitting them into contiguous components of numbers and  letters. E.g., 3.3.1pre5 is split into  [3, 3,  1,  pre,  5]. These  lists  are then  compared lexicographically  (from left to right).  
> Corresponding components a  and b  are compared as  follows.  

> If they are both numbers, integer  comparison is used. 
> If a is an empty string and b is a number, a is considered  less than b. 
> The special string component pre (for pre-release) is considered to  be less  than  other components.  
> String components  are considered less than number components. 
> Otherwise, they are compared   lexicographically  (i.e., using case-sensitive string comparison).

Reply:
Am I confused about this or does that not resolve the problem? "[String components] are compared lexicographocally". My understanding is that Nix will regard `0_0` as newer than `0.1`.

Reply:
"0_0" is [0, "_", 0] thus smaller than "0.0" but larger than "0"

Lexicographic comparisons applies to string components after splitting.

"0.pre" seems to be smaller than anything not starting with "0.pre.pre" (dot after zero is optional). I _hope_ that "0.pre.pre" upstream version will not be a frequent issue.

Reply:
Isn't `0.0` split into [`0`, `.`, `0`]? I didn't read anything special about dots on that man page.

Reply:
Ah, the man page does not mention this but both `.` and `-` are special-cased separators that don't become part of any component! I had assumed the `.` was part of the evaluation.

Reply:
holy cow. just use `sort -V`

```
printf "%s\n" 0.0 0_0 1.0 0.1 2.0 \
1.0-unstable-2023-02-04 1.0+unstable-2023-02-04 \
| sort -V

0.0
0.1
0_0
1.0
1.0+unstable-2023-02-04
1.0-unstable-2023-02-04
2.0
```

Reply:
Nix does not produce the same order, `pre` is special-cased to sort earlier than in `sort -V`, and this is not a bug but a reasonable decision, compatibility with which we want to keep to avoid globally-breaking changes

Reply:
aah, makes sense

> `builtins.sort (a: b: (builtins.compareVersions a b) < 0) [ "000" "00" "0_0" "0" "0.0" "1.0" "1.0pre" "1.0-pre" "1.0post" "1.0-post" "1.0-unstable-2023-02-04" "1.0-unstable.2023-02-04" ]`
>
> `[ "000" "00" "0" "0_0" "0.0" "1.0pre" "1.0-pre" "1.0" "1.0post" "1.0-post" "1.0-unstable-2023-02-04" "1.0-unstable.2023-02-04" ]`